api/client/lib/authn.go

```diff
@@ -0,0 +1,211 @@
+	// the responder should be called.
+	scheme() string
+	// authRespond, given the authentication header value associated with
+	// the scheme that it implements, can decide if the request should be
+	// retried.  If it returns true, then the request is retransmitted to
+	// the server, presumably because it has added an authentication header
+	// which it believes the server will accept.
+	authRespond(cli *Client, challenge string, req *http.Request) (bool, error)
+	// AuthCompleted, given a (possibly empty) WWW-Authenticate header from
+	// a successful response, should decide if the server's reply should be
+	// accepted.
+	authCompleted(cli *Client, challenge string, req *http.Request) (bool, error)
+}
+
+// authResponderCreator either creates a new authResponder, or returns nil
+type authResponderCreator func() authResponder
+
+var authResponderCreators = []authResponderCreator{}
+
+// registerAuthResponder registers a function which will be called at startup
+// to create an authResponder.
+func registerAuthResponder(arc authResponderCreator) {
+	authResponderCreators = append(authResponderCreators, arc)
+}
+
+// Run through all of the registered responder creators and build a map of
+// lists of responders.
+func createAuthResponders() *map[string]*[]authResponder {
+	ars := make(map[string]*[]authResponder)
+	for _, arc := range authResponderCreators {
+		responder := arc()
+		if responder != nil {
+			scheme := strings.ToLower(responder.scheme())
+			if ars[scheme] != nil {
+				slice := append(*(ars[scheme]), responder)
+				ars[scheme] = &slice
+			} else {
+				ars[scheme] = &[]authResponder{responder}
+			}
+		}
+	}
+	return &ars
+}
+
+// doer is just an interface wrapper that we use to be able to Do() using
+// either an http.Client or an httputil.Clientconn.
+type doer interface {
+	Do(req *http.Request) (resp *http.Response, err error)
+}
+
+// Run the Doer's Do() method, setting and saving cookies.  An http.Client
+// would handle this for us if we passed in a cookie jar when we initialized
+// it, but we'd still have to do the work for httputil.ClientConn, so we don't
+// set it for http.Clients to avoid duplicating cookies in requests.
+func (cli *Client) doWithCookies(doer doer, req *http.Request) (*http.Response, error) {
+	if localhost, ok := os.Hostname(); ok == nil && cli.proto == "unix" {
+		req.URL.Host = localhost
+	} else {
+		req.URL.Host = cli.addr
+	}
+	req.URL.Scheme = cli.scheme
+	if cli.jar != nil {
+		for _, cookie := range cli.jar.Cookies(req.URL) {
+			req.AddCookie(cookie)
+		}
+	}
+	resp, err := doer.Do(req)
+	if cli.jar != nil && resp != nil {
+		if cookies := resp.Cookies(); cookies != nil {
+			cli.jar.SetCookies(req.URL, cookies)
+		}
+	}
+	return resp, err
+}
+
+// doWithAuthn calls Do(), and handles any "unauthorized" errors
+// which it returns by retrying it with authentication.
+func (cli *Client) doWithAuthn(doer doer, req *http.Request, body []byte) (*http.Response, error) {
```

Comment:
I don't think we should have methods with 7 indentation levels and 3 nested for's. This is really hard to read and should be refactored somehow.